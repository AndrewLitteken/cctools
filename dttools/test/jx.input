{
"outfile":"results",
"infile":"mydata",
"a": true,
"b": false,
"f": 0.5,
"g": 3.141592654,
"x": 10,
"y": 20,
"list": [100,200,300],
"object": { "house": "home" }
}

#comment
10;
3847.576;
.5;
10.;
987;
12347812309487;

"hello";
"goodbye\n";
"\"quotes\"";
"tab\ttab\ttab\rreturn\nnewline\n";
"\\\\\\\\";

true;
false;
null;
true && true;
true && false;
false && true;
false && false;
true || true;
true || false;
false || true;
false || false;
!true;
!false;

10 == 20;
10 != 20;
10 < 20;
10 <= 20;
10 > 20;
10 >= 20;

10.5 == 20.5;
10.5 != 20.5;
10.5 < 20.5;
10.5 <= 20.5;
10.5 > 20.5;
10.5 >= 20.5;

"hello" == "goodbye";
"hello" != "goodbye";
"hello" <= "goodbye";
"hello" < "goodbye";
"hello" > "goodbye";
"hello" >= "goodbye";

"hello" == "hello";
"hello" != "hello";
"hello" <= "hello";
"hello" < "hello";
"hello" > "hello";
"hello" >= "hello";

a==b;
a!=b;
a<b;
a>b;
a<=b;
a>=b;

f==g;
f!=g;
f<g;
f>g;
f<=g;
f>=g;

x==y;
x!=y;
x<y;
x>y;
x<=y;
x>=y;

a+b;
a-b;
a*b;
a/b;
a%b;

f+g;
f-g;
f*g;
f/g;
f%g;

x+y;
x-y;
x*y;
x/y;
x%y;
x&&y;
x||y;

(x+y)*(f+g);
(10 + (20 * 30));
(10*(20+30));
(a==b) || ((f>g) && x<y);

#[10,-90.5,true,false,null,[-1,-2,-3]];
[10,-90.5,true,false,null,[-1,-2,-3]];

list[2];
object["house"];

{
"command":"grep English "+infile+" > "+outfile,
"inputs":["/usr/bin/grep",infile],
"outputs":[outfile],
"environment":{"PATH":"/usr/bin"},
"cores":1,
"memory":16,
"disk":1,
};

range(5);
range(3, 7);
range (7, 3);
range(-1, 10, 2);
range(1, 10, 0);
range(0, 5, -1);
range(5, 0, 1);
range(5, 0, -1);

format();
format("%%");
format("value: %i!!", x + y);
format("%i", x, y);
format("(%d, %i)", x, y);
format("%i");
format("%e", 1.2e-22);
format("%E", 6.02e23);
format("%f", 2.5);
format("%F", 3.14);
format("%g", 9.9e9);
format("%G", 2.11111e-12);
format("%b", false);
format("%s", "foo");

10 + [1];
"abc" + [2];
[] + [];
[] + [1];
[1] + [];
[1, 2] + [3];
[1, 2] + [] + [4, [5, 6]];
[10] == [10];
[10] != [10];
[1,2, 3] == [1,3,2];
[1,2,[3,4]] == [1,2,[3,4]];

Error{"source":"jx_eval","op":10+[1],"line":409,"file":"jx_eval.c","message":"mismatched types for operator","name":"TypeError"};

#end
