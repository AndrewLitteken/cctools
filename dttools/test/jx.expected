Enter context expression (or {} for an empty context):
Now enter expressions:
expression: 10
value:      10

expression: 3847.58
value:      3847.58

expression: 0.5
value:      0.5

expression: 10
value:      10

expression: 987
value:      987

expression: 12347812309487
value:      12347812309487

expression: "hello"
value:      "hello"

expression: "goodbye\n"
value:      "goodbye\n"

expression: "\"quotes\""
value:      "\"quotes\""

expression: "tab\ttab\ttab\rreturn\nnewline\n"
value:      "tab\ttab\ttab\rreturn\nnewline\n"

expression: "\\\\\\\\"
value:      "\\\\\\\\"

expression: true
value:      true

expression: false
value:      false

expression: null
value:      null

expression: true&&true
value:      true

expression: true&&false
value:      false

expression: false&&true
value:      false

expression: false&&false
value:      false

expression: true||true
value:      true

expression: true||false
value:      true

expression: false||true
value:      true

expression: false||false
value:      false

expression: !true
value:      false

expression: !false
value:      true

expression: 10==20
value:      false

expression: 10!=20
value:      true

expression: 10<20
value:      true

expression: 10<=20
value:      true

expression: 10>20
value:      false

expression: 10>=20
value:      false

expression: 10.5==20.5
value:      false

expression: 10.5!=20.5
value:      true

expression: 10.5<20.5
value:      true

expression: 10.5<=20.5
value:      true

expression: 10.5>20.5
value:      false

expression: 10.5>=20.5
value:      false

expression: "hello"=="goodbye"
value:      false

expression: "hello"!="goodbye"
value:      true

expression: "hello"<="goodbye"
value:      false

expression: "hello"<"goodbye"
value:      false

expression: "hello">"goodbye"
value:      true

expression: "hello">="goodbye"
value:      true

expression: "hello"=="hello"
value:      true

expression: "hello"!="hello"
value:      false

expression: "hello"<="hello"
value:      true

expression: "hello"<"hello"
value:      false

expression: "hello">"hello"
value:      false

expression: "hello">="hello"
value:      true

expression: a==b
value:      false

expression: a!=b
value:      true

expression: a<b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":72,"operator":true<false,"code":1}

expression: a>b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":73,"operator":true>false,"code":1}

expression: a<=b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":74,"operator":true<=false,"code":1}

expression: a>=b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":75,"operator":true>=false,"code":1}

expression: f==g
value:      false

expression: f!=g
value:      true

expression: f<g
value:      true

expression: f>g
value:      false

expression: f<=g
value:      true

expression: f>=g
value:      false

expression: x==y
value:      false

expression: x!=y
value:      true

expression: x<y
value:      true

expression: x>y
value:      false

expression: x<=y
value:      true

expression: x>=y
value:      false

expression: a+b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":91,"operator":true+false,"code":1}

expression: a-b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":92,"operator":true-false,"code":1}

expression: a*b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":93,"operator":true*false,"code":1}

expression: a/b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":94,"operator":true/false,"code":1}

expression: a%b
value:      Error{"source":"jx_eval","name":"unsupported operator","message":"unsupported operator on boolean","line":95,"operator":true%false,"code":1}

expression: f+g
value:      3.64159

expression: f-g
value:      -2.64159

expression: f*g
value:      1.5708

expression: f/g
value:      0.159155

expression: f%g
value:      0

expression: x+y
value:      30

expression: x-y
value:      -10

expression: x*y
value:      200

expression: x/y
value:      0

expression: x%y
value:      10

expression: x&&y
value:      0

expression: x||y
value:      30

expression: (x+y)*(f+g)
value:      109.248

expression: 10+20*30
value:      610

expression: 10*(20+30)
value:      500

expression: a==b||(f>g&&x<y)
value:      false

expression: [10,-90.5,true,false,null,[-1,-2,-3]]
value:      [10,-90.5,true,false,null,[-1,-2,-3]]

expression: list[2]
value:      300

expression: object["house"]
value:      "home"

expression: {"command":"grep English "+infile+" > "+outfile,"inputs":["/usr/bin/grep",infile],"outputs":[outfile],"environment":{"PATH":"/usr/bin"},"cores":1,"memory":16,"disk":1}
value:      {"command":"grep English mydata > results","inputs":["/usr/bin/grep","mydata"],"outputs":["results"],"environment":{"PATH":"/usr/bin"},"cores":1,"memory":16,"disk":1}

expression: range(5)
value:      [0,1,2,3,4]

expression: range(3,7)
value:      [3,4,5,6]

expression: range(7,3)
value:      []

expression: range(-1,10,2)
value:      [-1,1,3,5,7,9]

expression: range(1,10,0)
value:      Error{"source":"jx_eval","name":"invalid arguments","message":"step must be nonzero","function":range(1,10,0),"code":6}

expression: range(0,5,-1)
value:      []

expression: range(5,0,1)
value:      []

expression: range(5,0,-1)
value:      [5,4,3,2,1]

expression: format()
value:      Error{"source":"jx_eval","name":"invalid arguments","message":"invalid/missing format string","function":format(),"code":6}

expression: format("%%")
value:      "%"

expression: format("value: %i!!",x+y)
value:      "value: 30!!"

expression: format("%i",x,y)
value:      "10"

expression: format("(%d, %i)",x,y)
value:      "(10, 20)"

expression: format("%i")
value:      Error{"source":"jx_eval","name":"invalid arguments","message":"mismatched format specifier","function":format("%i"),"code":6}

expression: format("%e",1.2e-22)
value:      "1.200000e-22"

expression: format("%E",6.02e+23)
value:      "6.020000E+23"

expression: format("%f",2.5)
value:      "2.500000"

expression: format("%F",3.14)
value:      "3.140000"

expression: format("%g",9.9e+09)
value:      "9.9e+09"

expression: format("%G",2.11111e-12)
value:      "2.11111E-12"

expression: format("%b",false)
value:      "false"

expression: format("%s","foo")
value:      "foo"

expression: 10+[1]
value:      Error{"source":"jx_eval","name":"mismatched types","message":"mismatched types for operator","line":156,"operator":10+[1],"code":2}

expression: "abc"+[2]
value:      Error{"source":"jx_eval","name":"mismatched types","message":"mismatched types for operator","line":157,"operator":"abc"+[2],"code":2}

expression: []+[]
value:      []

expression: []+[1]
value:      [1]

expression: [1]+[]
value:      [1]

expression: [1,2]+[3]
value:      [1,2,3]

expression: [1,2]+[]+[4,[5,6]]
value:      [1,2,4,[5,6]]

expression: [10]==[10]
value:      true

expression: [10]!=[10]
value:      false

expression: [1,2,3]==[1,3,2]
value:      false

expression: [1,2,[3,4]]==[1,2,[3,4]]
value:      true

expression: Error{"source":"jx_eval","op":10+[1],"line":409,"file":"jx_eval.c","message":"mismatched types for operator","name":"TypeError"}
value:      Error{"source":"jx_eval","op":10+[1],"line":409,"file":"jx_eval.c","message":"mismatched types for operator","name":"TypeError"}

